### ISSUE TITLE
(auto context) Main-repo should not require git repo and local configs should be usable only locally

### ISSUE THREAD

**Issue by tuukkamustonen** [nosarthur/gita#179]
1) Why is there requirement for a "main-repo" to be a git repo? Can't I just do local config under folder X, which would contain actual git repositories?

Is it even a use case to have repositories under a repository (unless you're using submodule, subtree, or similar)?

2) Also, when you `cd ..` outside of the local config directory (I had it called `projects`):

```console
$ cd ..
$ gita ll
projects master _    ()
```

But then:

```console
$ gita ll projects
usage: gita ll [-h] [-C] [{}]
gita ll: error: argument group: invalid choice: 'projects' (choose from )
```

Should I be able to interact with the repo from outside the containing directory, or why does it get listed?

Also, in another main-repo:

```
$ gita ll
project1-repo   master      ()
...
projects        master _    ()
...
```

So, it shows listing of main-repos when inside another main-repo. This is likely unwanted behavior - I would expect main-repo (could it be simply called "local config"?) to only list repos under that main-repo. Shouldn't all main-repos be excluded from that listing?

**Comment by nosarthur** [comment]
I don't really use the main repo mechanism. I only use groups and context. Currently, the valid contexts are group names and `none`, which means no context (show all).

I can add another special context `auto`, which automatically sets the context depending on the working directory

**Comment by nosarthur** [comment]
`auto` is in 0.15.5

**Comment by tuukkamustonen** [comment]
Hmm, I'm not sure what it does. With:

```
temp/
    project1/
        repoA
    project2/
        repoA
        repoB
```

```
$ cd temp
$ rm -rf ~/.config/gita .gita/
$ gita add -a .
Found 4 new repo(s).
Created 3 new group(s).
$ gita ll
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
temp           master _    ()
$ gita context auto
$ gita ll
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
temp           master _    ()
```

I think here with `gita ll` and context set to `auto` the `temp` could be omitted, at least (as we are inside it)?

```
$ cd ..
$ gita ll
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
temp           master _    ()
```

Now it's showing the structure outside the directory we were in. With `auto`, shouldn't we get empty listing here (or just `temp` at maximum)?

**Comment by tuukkamustonen** [comment]
Also, if there's no real need for the "main repo" to be a git repo, why not rename it as "local config" and remove the requirement for git repository in the main level?

The difference between `gita add -m .` and `gita add -a .` is slightly confusing to me. Gita generates similar naming for both, but `-m` requires top level to be a git repo, and `-a` does not, plus it does auto-grouping. But then, `-m` config is local and stored to `.gita/` while `-a` is not. I think you could do some merging of the two features here, or at least polish the terminology and clarify the intent (you probably already saw https://github.com/nosarthur/gita/issues/180, too).

**Comment by nosarthur** [comment]
It's indeed confusing to have `-m` and `-a`. Since we have the `auto` context, maybe I will get rid of the 'main repo' mechanism and the local configurations.

This 'main repo'/local config idea seems to come from this issue
https://github.com/nosarthur/gita/issues/82

**Comment by tuukkamustonen** [comment]
Glancing through that ticket, I don't think the reporter was using a main _repo_. Rather, he wanted a local config for certain (any) folder. I believe there was never will to require the main directory to be a git repository (or maybe I missed that).

Anyway, what your typical user probably wants is `gita` to "activate" the local configuration, when you execute within a folder that contains such configuration (in `.gita/` directory).

So that with:

```
temp/
    project1/
        repoA
    project2/
        repoA
        repoB
```

```
$ rm -rf ~/.config/gita .gita/
$ gita add -a .
...
$ gita context auto
$ gita ll
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
$ cd project1
$ gita ll
project1/repoA master      ()  <-- currently shows all repos
$ cd ../project2
$ gita ll
project2/repoA master      ()  <-- currentl shows all repos
repoB          master      ()
$ cd ../..
$ gita ll
<nothing>  <-- currently shows all repos
```

What happens with current version is that full listing of repos is shown everywhere, in `temp`, in `project1`, in `project2` and in `..`. I think a local `.gita/` directory makes a lot of sense and looks nice, you don't need to put anything under `~`.

But then, sure you can skip local `.gita/` directories and have `auto` context interpret the situation accordingly. As long as the behavior is such that you get listed only the projects that exist in the directory or any of the descendants of it.

**Comment by nosarthur** [comment]
Your example snippet seems to have the correct behavior. Not sure if that's what you want, or what is shown on your terminal

**Comment by tuukkamustonen** [comment]
> Your example snippet seems to have the correct behavior. Not sure if that's what you want, or what is shown on your terminal

Yeah, I imiitated "what your typical user probably wants" so I adjusted the output to have the "correct" format. What actually gets printed is the same full:

```
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
```

...always.

> Is your temp also a git repo? If not, I think it would work correctly.

`temp` is/was not a git repo.

If you're saying that it _should_ work, maybe I made a mistake. Let me try again...

**Comment by nosarthur** [comment]
Sorry, you are right, it's not working, I only checked that the grouping is correct. It shouldn't be hard to fix this given the current setup...

**Comment by tuukkamustonen** [comment]
Just to clarify:

- When `temp` is NOT a git repo and I use `gita add -a .` and `gita context auto` --> the listing is always the same, no matter in which directory I am. Maybe that's intentional, I'm not sure?
- When `temp` IS a git repo and I use `gita add -m .` and `gita context auto` --> the listing is same for `temp` and it's sub-directories (`project1` and `project2`). The listing only shows `temp master _    ()` when I `cd ..` out of temp.

I think this might just get more confusing until you re-work/merge the `-m` and `-a` flags :) Take a look at https://github.com/nosarthur/gita/issues/178#issuecomment-911730250 - once/if you re-work those flags, I can try again then.

**Comment by nosarthur** [comment]
It'll get better if we get rid of `-m` completely, and set `auto` context. Whether there is a local config is an implementation detail that the user doesn't need to worry about.

Speed might be a concern for the current implementation (no local config files for `-a`), but we can tune it up later. 

The complication of local setup is that how to find the local setting. That might be the reason of the 'main repo' concept: 'main repo' paths are where the local settings are, thus `cwd` is only compared to them. But we can surely drop the requirement that local config can only exist at the root of a repo

**Comment by tuukkamustonen** [comment]
> It'll get better if we get rid of -m completely, and set auto context. Whether there is a local config is an implementation detail that the user doesn't need to worry about.

While that's true, it's just convenient to see the `.gita/` directory. It makes more obvious to the user where the configs are. And if you copy stuff around, your configs get copied too (unless you use absolute paths in there, then of course things would break when copying).

Though I agree, that it's kinda trivial. And if you want to be aware of the "local contexts" outside of their nesting directories, then you need at least some entry under `~` anyway.

Just thinking out loud - pick whatever style you think suits best!

**Comment by tuukkamustonen** [comment]
By the way, if the idea is to be aware of the "local" contexts outside of them, (and even if not) maybe add `gita cd <group/repo>` so quickly change directory to the target?

**Comment by nosarthur** [comment]
I use 

```
cd `gita ls <repo>`
```

Not sure if it's a bash grammar, same effect with `cd $(gita ls <repo>)`

It's not possible to do `gita cd <repo>`, see for example 
https://stackoverflow.com/questions/431684/equivalent-of-shell-cd-command-to-change-the-working-directory

> Changing the current working directory in a subprocess does not change the current working directory in the parent process. This is true of the Python interpreter as well. You cannot use os.chdir() to change the CWD of the calling process.

**Comment by tuukkamustonen** [comment]
Ah true, fair point. Thanks for tip!

**Comment by nosarthur** [comment]
I haven't clean up the tests yet, but it seems to be working: v0.15.6.2

**Comment by tuukkamustonen** [comment]
@nosarthur I see you've been working on it :)

Should I try a new version now, or are you expected to adjust the behavior, still?

**Comment by nosarthur** [comment]
Please try the latest version, it's working on my machine. Let me know if you find any issues

**Comment by tuukkamustonen** [comment]
On 0.15.7 and with my test setup of

```
temp/
    project1/
        repoA
    project2/
        repoA
        repoB
```

`temp` and `projectX` are NOT git repos. `repoA`s and `repoB` are.

Then:

```
$ gita add -a .
Found 3 new repo(s).
Created 3 new group(s).
$ gita ll
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
$ gita ll -g
temp:
temp-project2:
temp-project1:
```

So, `ll -g` is not working as it only shows groups and not their repos?

Continuing:

```
$ gita context auto
$ gita ll -g
temp:
   project1/repoA master      ()
   project2/repoA master      ()
   repoB          master      ()
```

With `auto` context it changes behavior of `ll -g`. Now it's missing other than "temp" groups (which is missing in `gita ll` listing btw.)?

```
$ cd ..
$ gita ll
project1/repoA master      ()
project2/repoA master      ()
repoB          master      ()
$ gita ll -g
temp:
temp-project2:
temp-project1:
```

I don't know, shouldn't it either show nothing or just "temp" above because we `cd ..` and are using `auto` context? Now it shows everything despite us being away from the containing directory.

I'm not sure of the design you went with here, so maybe I misunderstood something?

**Comment by nosarthur** [comment]
Thanks for the feed back

```
$ gita ll -g
temp:
temp-project2:
temp-project1:
```
This bug is fixed in 0.15.7.1

With auto context, the repos are restricted, thus the later behaviors are expected.

You can see the current context with `gita context`. When you are out of `temp`, there is no context set, thus all groups are shown. Does it make sense?
